src/unit.rs

```diff
@@ -163,12 +160,69 @@ impl Unit {
     }
 }
 
-/// Perform a connection. Used recursively for redirects.
+/// Perform a connection. Follows redirects.
 pub(crate) fn connect(
-    unit: Unit,
+    mut unit: Unit,
+    use_pooled: bool,
+    mut body: SizedReader,
+) -> Result<Response, Error> {
+    let mut previous = vec![];
+    loop {
+        let mut resp = connect_inner(&unit, use_pooled, body, &previous)?;
+        previous.push(unit.url.to_string());
+        resp.history = previous;
```

Comment:
I think this is slightly early to push the unit's URL onto `previous`. Doesn't this mean a resp with no redirects contains itself in its history? We should make sure we have tests that catch this, if it's the case (I might also be misunderstanding here).

Reply:
Ooh, excellent catch. Yes, this should only push the history if we're sure we're going to redirect (near the bottom, after calculating `new_method`.

That brings up another issue: I don't think the history tests actually test anything. Because they add items to the history manually, it doesn't matter what `connect` does, they'll pass no matter what.